### PR TITLE
Chat SDK

### PR DESCRIPTION
**Pull request by bloodbare**
### Description
Describe the proposed changes made in this PR.

### How was this PR tested?
Describe how you tested this PR.

### CODE REVIEW

File: nucliadb/nucliadb/search/predict.py [hunk 1]
```diff
@@ -138,10 +154,47 @@ async def send_feedback(
             )
         await self.check_response(resp, expected=204)
 
+    @predict_observer.wrap({"type": "chat"})
```
Comment (lferran) [inline nucliadb/nucliadb/search/predict.py:157]:
```suggestion
    @predict_observer.wrap({"type": "rephrase"})
```

File: nucliadb/nucliadb/search/api/v1/chat.py [hunk 2]
```diff
@@ -99,17 +99,16 @@ async def chat(
...
             user_id=x_nucliadb_user,
             retrieval=False,
         )
-        req.system = "You help creating new queries based on context"
 
         new_query_elements = []
-        _, generator = await predict.chat_query(kbid, req)
+        _, generator = await predict.rephrase_query(kbid, req)
         async for new_query_data in generator:
+            new_query = ""
```
Comment (jotare) [inline nucliadb/nucliadb/search/api/v1/chat.py:111]: What's the point of this?

File: nucliadb/nucliadb/search/predict.py (hunk 1 above)
Comment (jotare) [inline nucliadb/nucliadb/search/predict.py:157]: Maybe type rephrase?

File: nucliadb/nucliadb/search/predict.py [hunk 3]
```diff
@@ -138,10 +154,47 @@ async def send_feedback(
...
+        if self.onprem is False:
+            # Upload the payload
+            resp = await self.session.post(
+                url=f"{self.cluster_url}{PRIVATE_PREDICT}{REPHRASE}",
+                json=item.dict(),
+                headers={"X-STF-KBID": kbid},
+            )
+        else:
+            if self.nuclia_service_account is None:
+                error = "Nuclia Service account is not defined so could not retrieve vectors for the query"
```
Comment (lferran) [inline nucliadb/nucliadb/search/predict.py:172]:
```suggestion
                error = "Nuclia Service account is not defined so could not rephrase query"
```

File: nucliadb_sdk/nucliadb_sdk/tests/fixtures.py [hunk 4]
```diff
@@ -34,10 +38,18 @@
     "version": "latest",
     "env": {
         "NUCLIADB_DISABLE_TELEMETRY": "True",
+        "dummy_processing": "True",
         "max_receive_message_length": "40",
+        "TEST_SENTENCE_ENCODER": "multilingual-2023-02-21",
+        "TEST_RELATIONS": """{"tokens": [{"text": "Nuclia", "ner": "ORG"}]}""",
+        # "TEST_RELATIONS": """{"tokens": []}""",
```
Comment (jotare) [inline nucliadb_sdk/nucliadb_sdk/tests/fixtures.py:45]: Can be deleted?

File: nucliadb/nucliadb/search/predict.py [hunk 5]
```diff
@@ -138,10 +154,47 @@ async def send_feedback(
...
         self, kbid: str, item: ChatModel
     ) -> Tuple[str, AsyncIterator[bytes]]:
+        if self.dummy:
+            self.calls.append(item)
+
+            async def generate():
+                for i in [b"valid ", b"answer ", b" to"]:
+                    yield i
+
+            return ("00", generate())
```
Comment (lferran) [inline nucliadb/nucliadb/search/predict.py:196]: what is this "00" for? Do we have this documented somewhere?

File: nucliadb_sdk/nucliadb_sdk/client.py [hunk 6]
```diff
@@ -81,25 +87,25 @@ def __init__(
...
+        # self.url = url
 
         internal_hosts_set = all((writer_host, reader_host, search_host, train_host))
         url_set = bool(url)
 
-        if not (url_set or internal_hosts_set):
-            raise AttributeError("Either url or nucliadb services hosts must be set")
-
-        if url_set:
+        if url_set and url is not None:
```
Comment (lferran) [inline nucliadb_sdk/nucliadb_sdk/client.py:95]:
```suggestion
        if url_set:
```

File: nucliadb_sdk/nucliadb_sdk/client.py [hunk 7]
```diff
@@ -81,25 +87,25 @@ def __init__(
...
             self.url = url
-        elif internal_hosts_set:
+        elif internal_hosts_set and writer_host is not None:
```
Comment (lferran) [inline nucliadb_sdk/nucliadb_sdk/client.py:97]: I think this is redundant, as `writer_host` is being checked in `internal_hosts_set`.

So we were good with `elif internal_hosts_set`

File: nucliadb_sdk/nucliadb_sdk/find.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class FindResult:
+    inner_find_results: KnowledgeboxFindResults
+
+    def __init__(
+        self, inner_find_results: KnowledgeboxFindResults, client: NucliaDBClient
+    ):
+        self.inner_find_results = inner_find_results
+        self.client = client
```
Comment (lferran) [inline nucliadb_sdk/nucliadb_sdk/find.py:46]: Is the client needed? As far as I can see it is not used

File: nucliadb_sdk/nucliadb_sdk/knowledgebox.py [hunk 9]
```diff
@@ -336,6 +355,145 @@ async def async_search(
...
+        text: Optional[str] = None,
+        filter: Optional[List[Union[Label, str]]] = None,
+        vector: Optional[Union[ndarray, List[float]]] = None,
+        vectorset: Optional[str] = None,
+        min_score: Optional[float] = 0.0,
+    ):
+        result = self.client.find(
+            self.build_find_request(text, filter, vector, vectorset, min_score)
+        )
+        return FindResult(result, self.client)
```
Comment (lferran) [inline nucliadb_sdk/nucliadb_sdk/knowledgebox.py:369]: why returning the client instance in the find result?

File: nucliadb/nucliadb/search/predict.py (hunk 5 above)
Comment (bloodbare) [reply nucliadb/nucliadb/search/predict.py:196]: Its an ID from learning

File: nucliadb_sdk/nucliadb_sdk/client.py (hunk 7 above)
Comment (bloodbare) [reply nucliadb_sdk/nucliadb_sdk/client.py:97]: Its for typing

File: nucliadb/nucliadb/search/predict.py (hunk 5 above)
Comment (jotare) [reply nucliadb/nucliadb/search/predict.py:196]: Could you move it to a constant or something more documented? A string literal related with learning repo has no readable meaning for nucliadb

**Pull request merged**